org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinWhenStringFilter.java

```diff
@@ -0,0 +1,110 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.LabelNode;
+import org.objectweb.asm.tree.LookupSwitchInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TableSwitchInsnNode;
+
+/**
+ * Filters bytecode that Kotlin compiler generates for <code>when</code>
+ * expressions with a <code>String</code>.
+ */
+public final class KotlinWhenStringFilter implements IFilter {
+
+	public void filter(final MethodNode methodNode,
+			final IFilterContext context, final IFilterOutput output) {
+		final Matcher matcher = new Matcher();
+		for (AbstractInsnNode i = methodNode.instructions
+				.getFirst(); i != null; i = i.getNext()) {
+			matcher.match(i, output);
+		}
+	}
+
+	private static class Matcher extends AbstractMatcher {
+		public void match(final AbstractInsnNode start,
+				final IFilterOutput output) {
+
+			cursor = start;
+
+			nextIsVar(Opcodes.ASTORE, "s");
+			nextIsVar(Opcodes.ALOAD, "s");
+			nextIsInvokeVirtual("java/lang/String", "hashCode");
+			nextIsSwitch();
+			if (cursor == null) {
+				return;
+			}
+
+			final AbstractInsnNode s = cursor;
+			final int hashCodes;
+			final LabelNode defaultLabel;
+			if (s.getOpcode() == Opcodes.LOOKUPSWITCH) {
+				final LookupSwitchInsnNode lookupSwitch = (LookupSwitchInsnNode) cursor;
+				defaultLabel = lookupSwitch.dflt;
+				hashCodes = lookupSwitch.labels.size();
+			} else {
+				final TableSwitchInsnNode tableSwitch = (TableSwitchInsnNode) cursor;
+				defaultLabel = tableSwitch.dflt;
+				hashCodes = tableSwitch.labels.size();
+			}
+
+			final Set<AbstractInsnNode> replacements = new HashSet<AbstractInsnNode>();
+			replacements.add(instructionAfterLabel(defaultLabel));
+
+			for (int i = 0; i < hashCodes; i++) {
+				while (true) {
+					nextIsVar(Opcodes.ALOAD, "s");
+					nextIs(Opcodes.LDC);
+					nextIsInvokeVirtual("java/lang/String", "equals");
+					// jump to next comparison or default case
+					nextIs(Opcodes.IFEQ);
+					final JumpInsnNode jump = (JumpInsnNode) cursor;
+					// jump to case
+					nextIs(Opcodes.GOTO);
+					if (cursor == null) {
+						return;
+					}
+
+					replacements.add(instructionAfterLabel(
+							((JumpInsnNode) cursor).label));
+
+					if (jump.label == defaultLabel) {
+						// end of comparisons for same hashCode
+						break;
+					}
+				}
+			}
+
+			output.ignore(s.getNext(), cursor);
+			output.replaceBranches(s, replacements);
+		}
+	}
+
+	private static AbstractInsnNode instructionAfterLabel(
```

Comment:
Idea: Make a variant of `skipNonOpcodes` in base class public:

    private static AbstractInsnNode skipNonOpcodes(AbstractInsnNode start)